### ISSUE TITLE
ticket hook order upon creation

### ISSUE THREAD

**Issue by glebsts** [glpi-project/glpi#12627]
### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

9.5.5

### Bug description

I added hooks in my plugin for `add_ticket` and `update_ticket` just to log some events, and found following:
I create ticket setting both "Requester" and "Assigned To" to some users (not myself, if it matters).
I have added stacktrace output to [QueuedNotification#forceSendFor](https://github.com/glpi-project/glpi/blob/9.5/bugfixes/inc/queuednotification.class.php#L615). I get following sequence:
```
#1  QueuedNotification::forceSendFor(Ticket, 117) called at [glpi/inc/commondbtm.class.php:1604]
#2  CommonDBTM->update(Array ([id] => 117,[takeintoaccount_delay_stat] => 1,[_disablenotif] => 1)) called at [glpi/inc/ticket.class.php:2335]
#3  Ticket->updateDateMod(117, ) called at [glpi/inc/commonitilactor.class.php:382]
#4  CommonITILActor->post_addItem() called at [glpi/inc/ticket_user.class.php:59]
#5  Ticket_User->post_addItem() called at [glpi/inc/commondbtm.class.php:1173]
#6  CommonDBTM->add(Array ([users_id] => 4,[type] => 2,[tickets_id] => 117,[_do_not_compute_takeintoaccount] => ,[_from_object] => 1)) called at [glpi/inc/commonitilobject.class.php:1926]
#7  CommonITILObject->post_addItem() called at [glpi/inc/ticket.class.php:1926]
#8  Ticket->post_addItem() called at [glpi/inc/commondbtm.class.php:1173]
#9  CommonDBTM->add(....)
```

and later 
```
#1  QueuedNotification::forceSendFor(Ticket, 117) called at [glpi/inc/commondbtm.class.php:1214]
#2  CommonDBTM->add(...)
```

Line numbers have not original offset, as I added my debug code there, but here are links to github:
Right after `add`:
https://github.com/glpi-project/glpi/blob/9.5/bugfixes/inc/commondbtm.class.php#L1177 leads us toadding new ticket user and through that to bumping ticket mod date, as can be seen from first stacktrace.
and later 
https://github.com/glpi-project/glpi/blob/9.5/bugfixes/inc/commondbtm.class.php#L1219

And it results in hook calling disorder - first `update_ticket`, then `add_ticket`.
That seems like non-logical behavior. It kinda makes sense that when something is being notified about new ticket, it could also get info about user that ticket was AssignedTo (`type = 2` in trace args referse to `CommonITILActor::ASSIGN`). 
But it should not trigger ticket update before triggering ticket add event. While `Ticket_User` is separate entity, semantically ticket is created coupld with assigned user, for creator it is single item - ticket+assignee. 
Imo it would make sense to first porocess `add_ticket` hooks even without assignee created, and then run `update_ticket` with information about ticket+assignee. 
Thinking again, why should user addition _together_ with ticket creation modify ticket mod timestamp? As I said, for creating user ticket creation is single event in single moment of time.

What would be your recommended way of handling those events in hooks? Hook events contains
`"updates":["takeintoaccount_delay_stat","date_mod"]` info, should it filter out something?

Thank you.

### Steps To reproduce

1. add some plugin with hooks to `add_ticket` and `update_ticket`
2. created new ticket with assignee picked upon filling ticket content
3. watch hook firing order

### Your GLPI setup information

GLPI 9.5.5 ( => glpi)
Installation mode: TARBALL
PHP 7.4.3

**Comment by cedric-anne** [comment]
Hi,

Changing all of this is not so easy, at it may have many side effects, for instance on notification sent when a ticket is created. Also, we cannot changing this in bugfixes versions, to ensure that we will not break existing plugins. So, if any change is done on this, it cannot be done before GLPI 10.1.0.

Regards

**Comment by glebsts** [comment]
any recommendations to detect such sequences? Can I rely on only mod time field presence in updated fileds list? (next to takeintoaccount field)

**Comment by cedric-anne** [comment]
I may be wrong, but I think the `_from_object` element will be present in `$this->input` in such cases.

**Comment by glebsts** [comment]
sorry, was not able to do things for few days. Well, I checked not `_from_object` (which is mentioned in stacktrace above), but it doesn't seem to be propagated into hook input :(
But I did a comparison of inputs for `update` and `add` events. "Add" has much more fields, but nothing that I could rely on yet. 
Also mind this type misalign for `takeintoaccount_delay_stat`, just noticed.
![image](https://user-images.githubusercontent.com/322967/190633422-a3634cab-6b9d-40f5-96c6-374bb7145814.png)


Did also comparison of "add" input with "real update" input triggered by editing ticket body.
Main indicator would be
```
    "updates": [
        "content",
        "date_mod"
    ],
```
Many diffs are about `"NULL"`vs `null` in timestamps, for example `closedate`, `solvedate` btw. Screenshot would be too long to attach.

**To summarize**, there is no good way to check in hook if update was triggered by adding related entity or not. 
One can run euristics on `updates` array to see if only `date_mod` and `takeintoaccount_delay_stat` fields were modifed (I have no clue what this last field is, so I assume it is not important for me), and through that see if update has any important info for your plugin..
